File: app/src/main/java/awais/instagrabber/repositories/DirectMessagesRepository.java
package awais.instagrabber.repositories;

import java.util.Map;

import awais.instagrabber.repositories.responses.directmessages.DirectBadgeCount;
import awais.instagrabber.repositories.responses.directmessages.DirectInboxResponse;
import awais.instagrabber.repositories.responses.directmessages.DirectThreadBroadcastResponse;
import awais.instagrabber.repositories.responses.directmessages.DirectThreadDetailsChangeResponse;
import awais.instagrabber.repositories.responses.directmessages.DirectThreadFeedResponse;
import retrofit2.Call;
import retrofit2.http.FieldMap;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.QueryMap;

public interface DirectMessagesRepository {

    @GET("/api/v1/direct_v2/inbox/")
    Call<DirectInboxResponse> fetchInbox(@QueryMap Map<String, Object> queryMap);

    @GET("/api/v1/direct_v2/threads/{threadId}/")
    Call<DirectThreadFeedResponse> fetchThread(@Path("threadId") String threadId,
                                               @QueryMap Map<String, Object> queryMap);

    @GET("/api/v1/direct_v2/get_badge_count/")
    Call<DirectBadgeCount> fetchUnseenCount();

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/broadcast/{item}/")
    Call<DirectThreadBroadcastResponse> broadcast(@Path("item") String item,
                                                  @FieldMap final Map<String, String> signedForm);

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/add_user/")
    Call<DirectThreadDetailsChangeResponse> addUsers(@Path("threadId") String threadId,
                                                     @FieldMap final Map<String, String> form);

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/remove_users/")
    Call<String> removeUsers(@Path("threadId") String threadId,
                             @FieldMap final Map<String, String> form);

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/update_title/")
    Call<DirectThreadDetailsChangeResponse> updateTitle(@Path("threadId") String threadId,
                                                        @FieldMap final Map<String, String> form);

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/add_admins/")
    Call<String> addAdmins(@Path("threadId") String threadId,
                           @FieldMap final Map<String, String> form);

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/remove_admins/")
    Call<String> removeAdmins(@Path("threadId") String threadId,
                              @FieldMap final Map<String, String> form);

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/items/{itemId}/delete/")
    Call<String> deleteItem(@Path("threadId") String threadId,
                            @Path("itemId") String itemId,
                            @FieldMap final Map<String, String> form);
}

File: app/src/main/java/awais/instagrabber/repositories/DirectMessagesRepository.kt
package awais.instagrabber.repositories

import awais.instagrabber.repositories.responses.directmessages.*
import retrofit2.Call
import retrofit2.http.*

interface DirectMessagesRepository {
    @GET("/api/v1/direct_v2/inbox/")
    suspend fun fetchInbox(@QueryMap queryMap: Map<String, String>): DirectInboxResponse

    @GET("/api/v1/direct_v2/pending_inbox/")
    suspend fun fetchPendingInbox(@QueryMap queryMap: Map<String, String>): DirectInboxResponse

    @GET("/api/v1/direct_v2/threads/{threadId}/")
    suspend fun fetchThread(
        @Path("threadId") threadId: String,
        @QueryMap queryMap: Map<String, String>,
    ): DirectThreadFeedResponse

    @GET("/api/v1/direct_v2/get_badge_count/?no_raven=1")
    suspend fun fetchUnseenCount(): DirectBadgeCount

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/broadcast/{item}/")
    suspend fun broadcast(
        @Path("item") item: String,
        @FieldMap signedForm: Map<String, String>,
    ): DirectThreadBroadcastResponse

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/add_user/")
    fun addUsers(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/remove_users/")
    fun removeUsers(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/update_title/")
    fun updateTitle(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/add_admins/")
    fun addAdmins(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/remove_admins/")
    fun removeAdmins(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/items/{itemId}/delete/")
    fun deleteItem(
        @Path("threadId") threadId: String,
        @Path("itemId") itemId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @GET("/api/v1/direct_v2/ranked_recipients/")
    fun rankedRecipients(@QueryMap queryMap: Map<String, String>): Call<RankedRecipientsResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/broadcast/forward/")
    fun forward(@FieldMap form: Map<String, String>): Call<DirectThreadBroadcastResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/create_group_thread/")
    fun createThread(@FieldMap signedForm: Map<String, String>): Call<DirectThread?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/mute/")
    fun mute(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/unmute/")
    fun unmute(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/mute_mentions/")
    fun muteMentions(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String?>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/unmute_mentions/")
    fun unmuteMentions(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @GET("/api/v1/direct_v2/threads/{threadId}/participant_requests/")
    fun participantRequests(
        @Path("threadId") threadId: String,
        @Query("page_size") pageSize: Int,
        @Query("cursor") cursor: String?,
    ): Call<DirectThreadParticipantRequestsResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/approve_participant_requests/")
    fun approveParticipantRequests(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/deny_participant_requests/")
    fun declineParticipantRequests(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/approval_required_for_new_members/")
    fun approvalRequired(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/approval_not_required_for_new_members/")
    fun approvalNotRequired(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/leave/")
    fun leave(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/remove_all_users/")
    fun end(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectThreadDetailsChangeResponse?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/approve/")
    fun approveRequest(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/decline/")
    fun declineRequest(
        @Path("threadId") threadId: String,
        @FieldMap form: Map<String, String>,
    ): Call<String?>

    @FormUrlEncoded
    @POST("/api/v1/direct_v2/threads/{threadId}/items/{itemId}/seen/")
    fun markItemSeen(
        @Path("threadId") threadId: String,
        @Path("itemId") itemId: String,
        @FieldMap form: Map<String, String>,
    ): Call<DirectItemSeenResponse?>
}